realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsRemoteMongoCollection.java

```diff
@@ -38,12 +65,231 @@ public long getNativeFinalizerPtr() {
+        nativeInsertOne(nativePtr, encodedDocument, callback);
+        return ResultHandler.handleResult(success, error);
+    }
+
+    public RemoteInsertManyResult insertMany(final List<? extends DocumentT> documents) {
+        AtomicReference<RemoteInsertManyResult> success = new AtomicReference<>(null);
+        AtomicReference<ObjectServerError> error = new AtomicReference<>(null);
+        OsJNIResultCallback<RemoteInsertManyResult> callback = new OsJNIResultCallback<RemoteInsertManyResult>(success, error) {
+            @Override
+            protected RemoteInsertManyResult mapSuccess(Object result) {
+                Object[] objects = (Object[]) result;
+                Map<Long, BsonValue> insertedIdsMap = new HashMap<>();
+                for (int i = 0; i < objects.length; i++) {
+                    ObjectId objectId = (ObjectId) objects[i];
+                    BsonValue bsonObjectId = new BsonObjectId(objectId);
+                    insertedIdsMap.put((long) i, bsonObjectId);
+                }
+                return new RemoteInsertManyResult(insertedIdsMap);
+            }
+        };
+
+        String encodedDocumentArray = JniBsonProtocol.encode(documents, codecRegistry);
+        nativeInsertMany(nativePtr, encodedDocumentArray, callback);
+        return ResultHandler.handleResult(success, error);
+    }
+
+    public RemoteDeleteResult deleteOne(final Bson filter) {
+        AtomicReference<RemoteDeleteResult> success = new AtomicReference<>(null);
+        AtomicReference<ObjectServerError> error = new AtomicReference<>(null);
+        OsJNIResultCallback<RemoteDeleteResult> callback = new OsJNIResultCallback<RemoteDeleteResult>(success, error) {
+            @Override
+            protected RemoteDeleteResult mapSuccess(Object result) {
+                return new RemoteDeleteResult((Long) result);
+            }
+        };
+
+        String jsonDocument = JniBsonProtocol.encode(filter, codecRegistry);
+        nativeDeleteOne(nativePtr, jsonDocument, callback);
+        return ResultHandler.handleResult(success, error);
+    }
+
+    public RemoteDeleteResult deleteMany(final Bson filter) {
+        AtomicReference<RemoteDeleteResult> success = new AtomicReference<>(null);
+        AtomicReference<ObjectServerError> error = new AtomicReference<>(null);
+        OsJNIResultCallback<RemoteDeleteResult> callback = new OsJNIResultCallback<RemoteDeleteResult>(success, error) {
+            @Override
+            protected RemoteDeleteResult mapSuccess(Object result) {
+                return new RemoteDeleteResult((Long) result);
+            }
+        };
+
+        String jsonDocument = JniBsonProtocol.encode(filter, codecRegistry);
+        nativeDeleteMany(nativePtr, jsonDocument, callback);
+        return ResultHandler.handleResult(success, error);
+    }
+
+    private Object findOneSuccessMapper(@Nullable Object result, Class<?> resultClass) {
+        if (result == null) {
```

Comment:
As discussed be aware of the special `BsonNull` type. Do not know whether it is used in the OS layer or we can rely on standard null pointers.

Reply:
I've done a quick search and it doesn't seem to exist in the object server. Null is returned in some operations from the Stitch SDK, so my guess to stick to that for ensuring continuity.